File: barrierclass.cpp
#include <stdio.h>
#include <semaphore.h>
#include <thread>
#include <vector>
#include "barrierclass.h"

    Barrier::Barrier(int n) : n(n), count(0) {
        sem_init(&this->mutex, 0, 1);
        sem_init(&this->turnstile, 0, 0);
        sem_init(&this->turnstile2, 0, 0);
    }

    Barrier::~Barrier() {
        sem_destroy(&this->mutex);
        sem_destroy(&this->turnstile);
        sem_destroy(&this->turnstile2);
    }

    void Barrier::phase1() {
        sem_wait(&this->mutex);
        this->count += 1;

        printf("count before: %d\n", this->count);

        if (this->count == this->n) {
            for (int i = 0; i < this->n; ++i) {
                sem_post(&this->turnstile);
            }
        }

        sem_post(&this->mutex);

        sem_wait(&this->turnstile);
    }

    void Barrier::phase2() {
        sem_wait(&this->mutex);
        this->count -= 1;

        printf("count after: %d\n", this->count);

        if (this->count == 0) {
            for (int i = 0; i < this->n; ++i) {
                sem_post(&this->turnstile2);
            }
        }

        sem_post(&this->mutex);

        sem_wait(&this->turnstile2);
    }

    void Barrier::wait() {
        this->phase1();
        this->phase2();
    }

File: python/barrierclass.py
from threading import *

class Barrier:
    def __init__(self, n):
        self.n = n
        self.count = 0
        self.mutex = Semaphore(1)
        self.turnstile = Semaphore(0)
        self.turnstile2 = Semaphore(0)

    def phase1(self):
        self.mutex.acquire()
        self.count += 1

        print(f"count before: {self.count}")

        if self.count == self.n:
            self.turnstile.release(self.n)
        
        self.mutex.release()

        self.turnstile.acquire()

    def phase2(self):
        self.mutex.acquire()
        self.count -= 1

        print(f"count after: {self.count}")

        if self.count == 0:
            self.turnstile2.release(self.n)

        self.mutex.release()

        self.turnstile2.acquire()

    def wait(self):
        self.phase1()
        self.phase2()
        